File: 0962/solution.py
class Solution:
    def maxWidthRamp(self, nums: list[int]) -> int:
        stack = []
        for i in range(len(nums)):
            if not stack or nums[i] < nums[stack[-1]]:
                stack.append(i)
        res = 0
        for i in range(len(nums) - 1, -1, -1):
            while stack and nums[i] >= nums[stack[-1]]:
                res = max(res, i - stack[-1])
                stack.pop()
        return res

File: 0962/solution.java
class Solution {
   public int maxWidthRamp(int[] nums) {
      Stack<Integer> stack = new Stack<>();
      for (int i = 0; i < nums.length; i++) {
         if (stack.isEmpty() || nums[stack.getLast()] > nums[i]) {
            stack.push(i);
         }
      }
      int res = 0;
      for (int i = nums.length - 1; i >= 0; i--) {
         while (!stack.isEmpty() && nums[stack.getLast()] <= nums[i]) {
            res = Math.max(res, i - stack.getLast());
            stack.pop();
         }
         if (stack.isEmpty()) {
            return res;
         }
      }
      return res;
   }
}

// time O(n)
// space O(n)
